### PR TITLE
Add browser polyfils for Node.js modules (webpack 5 backwards compat)

### LINKED ISSUE #15948: Polyfill common Node.js packages for webpack 5
Webpack 5 stopped polyfilling Node.js packages, however, their usage is prolific throughout the npm ecosystem.

We should provide default polyfills for the most common usages:

- [x] Add `process` and `buffer`: https://github.com/vercel/next.js/pull/15499
- [ ] [`path`](https://github.com/vercel/next.js/issues/13341#issuecomment-669937493)

### PR DESCRIPTION
**Pull request by timneutkens**
This adds the following Node.js core polyfills only when the import is used:

- `path`
- `stream`
- `vm`
- `crypto`
- `buffer`

Fixes #15948

We'll have a separate issue about adding warnings for the usage of these modules in the browser, some polyfills like crypto are quite heavy and generally not needed for most applications (included accidentally through node_modules).

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| buildDuration | 13.1s | 13.8s | ⚠️ +689ms |
| nodeModulesSize | 65.6 MB | 65.7 MB | ⚠️ +194 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.37 | 2.422 | ⚠️ +0.05 |
| / avg req/sec | 1054.92 | 1032.31 | ⚠️ -22.61 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.428 | 1.442 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1750.13 | 1733.28 | ⚠️ -16.85 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..ec1a.js gzip | 9.99 kB | 9.99 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-008324e..83e7.js gzip | 6.74 kB | 6.74 kB | ✓ |
| webpack-ccf5..276a.js gzip | 751 B | 751 B | ✓ |
| Overall change | 56.6 kB | 56.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 5.89 kB | 5.89 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-c621fe2..dule.js gzip | 5.81 kB | 5.81 kB | ✓ |
| webpack-10c7..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 51.6 kB | 51.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| polyfills-75..1629.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-874bd8a..0103.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-fa39c..ec40.js gzip | 3.45 kB | 3.45 kB | ✓ |
| hooks-585f07..95a3.js gzip | 887 B | 887 B | ✓ |
| index-c7b63f..fc02.js gzip | 227 B | 227 B | ✓ |
| link-f4d2979..e57b.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..ebc7.js gzip | 284 B | 284 B | ✓ |
| withRouter-2..db68.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.71 kB | 7.71 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-97e743e..dule.js gzip | 626 B | 626 B | ✓ |
| _error-b4004..dule.js gzip | 2.3 kB | 2.3 kB | ✓ |
| hooks-696209..dule.js gzip | 387 B | 387 B | ✓ |
| index-a4dd74..dule.js gzip | 226 B | 226 B | ✓ |
| link-653c74f..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-1..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.37 kB | 5.37 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _buildManifest.js gzip | 273 B | 273 B | ✓ |
| _buildManife..dule.js gzip | 280 B | 280 B | ✓ |
| Overall change | 553 B | 553 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| index.html gzip | 948 B | 948 B | ✓ |
| link.html gzip | 954 B | 954 B | ✓ |
| withRouter.html gzip | 941 B | 941 B | ✓ |
| Overall change | 2.84 kB | 2.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| buildDuration | 14.5s | 14.4s | -15ms |
| nodeModulesSize | 65.6 MB | 65.7 MB | ⚠️ +194 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..ec1a.js gzip | 9.99 kB | 9.99 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-008324e..83e7.js gzip | 6.74 kB | 6.74 kB | ✓ |
| webpack-ccf5..276a.js gzip | 751 B | 751 B | ✓ |
| Overall change | 56.6 kB | 56.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 5.89 kB | 5.89 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-c621fe2..dule.js gzip | 5.81 kB | 5.81 kB | ✓ |
| webpack-10c7..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 51.6 kB | 51.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| polyfills-75..1629.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-874bd8a..0103.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-fa39c..ec40.js gzip | 3.45 kB | 3.45 kB | ✓ |
| hooks-585f07..95a3.js gzip | 887 B | 887 B | ✓ |
| index-c7b63f..fc02.js gzip | 227 B | 227 B | ✓ |
| link-f4d2979..e57b.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..ebc7.js gzip | 284 B | 284 B | ✓ |
| withRouter-2..db68.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.71 kB | 7.71 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-97e743e..dule.js gzip | 626 B | 626 B | ✓ |
| _error-b4004..dule.js gzip | 2.3 kB | 2.3 kB | ✓ |
| hooks-696209..dule.js gzip | 387 B | 387 B | ✓ |
| index-a4dd74..dule.js gzip | 226 B | 226 B | ✓ |
| link-653c74f..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-1..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.37 kB | 5.37 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _buildManifest.js gzip | 273 B | 273 B | ✓ |
| _buildManife..dule.js gzip | 280 B | 280 B | ✓ |
| Overall change | 553 B | 553 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _error.js | 1.03 MB | 1.03 MB | ✓ |
| 404.html | 4.18 kB | 4.18 kB | ✓ |
| hooks.html | 3.82 kB | 3.82 kB | ✓ |
| index.js | 1.03 MB | 1.03 MB | ✓ |
| link.js | 1.07 MB | 1.07 MB | ✓ |
| routerDirect.js | 1.06 MB | 1.06 MB | ✓ |
| withRouter.js | 1.06 MB | 1.06 MB | ✓ |
| Overall change | 5.26 MB | 5.26 MB | ✓ |

</details>

</details>
Commit: 173e1c0ce3094d1bcf90eff7e79363ac3d075ad4

**Review by ijjk — APPROVED** [review]
LGTM

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| buildDuration | 12.4s | 11.7s | -712ms |
| nodeModulesSize | 65.6 MB | 65.7 MB | ⚠️ +194 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.183 | 2.265 | ⚠️ +0.08 |
| / avg req/sec | 1145.41 | 1103.73 | ⚠️ -41.68 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.368 | 1.311 | -0.06 |
| /error-in-render avg req/sec | 1827.98 | 1906.55 | +78.57 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..ec1a.js gzip | 9.99 kB | 9.99 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-008324e..83e7.js gzip | 6.74 kB | 6.74 kB | ✓ |
| webpack-ccf5..276a.js gzip | 751 B | 751 B | ✓ |
| Overall change | 56.6 kB | 56.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 5.89 kB | 5.89 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-c621fe2..dule.js gzip | 5.81 kB | 5.81 kB | ✓ |
| webpack-10c7..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 51.6 kB | 51.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| polyfills-75..1629.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-874bd8a..0103.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-fa39c..ec40.js gzip | 3.45 kB | 3.45 kB | ✓ |
| hooks-585f07..95a3.js gzip | 887 B | 887 B | ✓ |
| index-c7b63f..fc02.js gzip | 227 B | 227 B | ✓ |
| link-f4d2979..e57b.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..ebc7.js gzip | 284 B | 284 B | ✓ |
| withRouter-2..db68.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.71 kB | 7.71 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-97e743e..dule.js gzip | 626 B | 626 B | ✓ |
| _error-b4004..dule.js gzip | 2.3 kB | 2.3 kB | ✓ |
| hooks-696209..dule.js gzip | 387 B | 387 B | ✓ |
| index-a4dd74..dule.js gzip | 226 B | 226 B | ✓ |
| link-653c74f..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-1..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.37 kB | 5.37 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _buildManifest.js gzip | 273 B | 273 B | ✓ |
| _buildManife..dule.js gzip | 280 B | 280 B | ✓ |
| Overall change | 553 B | 553 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| index.html gzip | 948 B | 948 B | ✓ |
| link.html gzip | 954 B | 954 B | ✓ |
| withRouter.html gzip | 941 B | 941 B | ✓ |
| Overall change | 2.84 kB | 2.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| buildDuration | 12.7s | 13.3s | ⚠️ +598ms |
| nodeModulesSize | 65.6 MB | 65.7 MB | ⚠️ +194 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..ec1a.js gzip | 9.99 kB | 9.99 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-008324e..83e7.js gzip | 6.74 kB | 6.74 kB | ✓ |
| webpack-ccf5..276a.js gzip | 751 B | 751 B | ✓ |
| Overall change | 56.6 kB | 56.6 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| 677f882d2ed8..dule.js gzip | 5.89 kB | 5.89 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| main-c621fe2..dule.js gzip | 5.81 kB | 5.81 kB | ✓ |
| webpack-10c7..dule.js gzip | 751 B | 751 B | ✓ |
| Overall change | 51.6 kB | 51.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| polyfills-75..1629.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-874bd8a..0103.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-fa39c..ec40.js gzip | 3.45 kB | 3.45 kB | ✓ |
| hooks-585f07..95a3.js gzip | 887 B | 887 B | ✓ |
| index-c7b63f..fc02.js gzip | 227 B | 227 B | ✓ |
| link-f4d2979..e57b.js gzip | 1.29 kB | 1.29 kB | ✓ |
| routerDirect..ebc7.js gzip | 284 B | 284 B | ✓ |
| withRouter-2..db68.js gzip | 284 B | 284 B | ✓ |
| Overall change | 7.71 kB | 7.71 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _app-97e743e..dule.js gzip | 626 B | 626 B | ✓ |
| _error-b4004..dule.js gzip | 2.3 kB | 2.3 kB | ✓ |
| hooks-696209..dule.js gzip | 387 B | 387 B | ✓ |
| index-a4dd74..dule.js gzip | 226 B | 226 B | ✓ |
| link-653c74f..dule.js gzip | 1.26 kB | 1.26 kB | ✓ |
| routerDirect..dule.js gzip | 284 B | 284 B | ✓ |
| withRouter-1..dule.js gzip | 282 B | 282 B | ✓ |
| Overall change | 5.37 kB | 5.37 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _buildManifest.js gzip | 273 B | 273 B | ✓ |
| _buildManife..dule.js gzip | 280 B | 280 B | ✓ |
| Overall change | 553 B | 553 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/webpack-polyfills | Change |
| - | - | - | - |
| _error.js | 1.03 MB | 1.03 MB | ✓ |
| 404.html | 4.18 kB | 4.18 kB | ✓ |
| hooks.html | 3.82 kB | 3.82 kB | ✓ |
| index.js | 1.03 MB | 1.03 MB | ✓ |
| link.js | 1.07 MB | 1.07 MB | ✓ |
| routerDirect.js | 1.06 MB | 1.06 MB | ✓ |
| withRouter.js | 1.06 MB | 1.06 MB | ✓ |
| Overall change | 5.26 MB | 5.26 MB | ✓ |

</details>

</details>
Commit: 9db55935d94c633521f3c8f9797bf6cace19fb4a

File: packages/next/build/webpack-config.ts [hunk 1]
```diff
@@ -356,7 +356,16 @@ export default async function getBaseWebpackConfig(
...
       'next/dynamic': 'next/dist/next-server/lib/dynamic.js',
-      next: NEXT_PROJECT_ROOT,
+      ...(isServer
+        ? {}
+        : {
+            stream: 'stream-browserify',
+            path: 'path-browserify',
+            crypto: 'crypto-browserify',
+            buffer: 'buffer',
+            vm: 'vm-browserify',
```
Comment (merceyz) [inline packages/next/build/webpack-config.ts:366]: @timneutkens These should go through`require.resolve`

Comment (timneutkens) [reply packages/next/build/webpack-config.ts:366]: Great point, forgot those don't resolve automatically. Opened #16033 with you as co-author.

**Pull request merged**